src/core/raster/qgsrasterlayer.cpp

```diff
@@ -611,7 +611,8 @@ void QgsRasterLayer::setDataProvider( QString const &provider, const QgsDataProv
   QgsDebugMsgLevel( QStringLiteral( "Entered" ), 4 );
   setValid( false ); // assume the layer is invalid until we determine otherwise
 
-  mPipe->remove( mDataProvider ); // deletes if exists
+  // deletes pipe elements (including data provider)
+  mPipe = std::make_unique< QgsRasterPipe >();
```

Comment:
Doesn't this lose the renderer settings?

Reply:
It's already being reset around like 727, and that's partly where the problem is. Calling mPipe->set( rendererPtr ) to replace a renderer that's already in the pipe leads to the issue described above whereas mDataProvider becomes unaccessible and calls to any of its functions lead to a segfault.